src/bounded_stack.ml

```diff
@@ -0,0 +1,208 @@
+type 'a node = Nil | Cons of { value : 'a; tail : 'a node; capacity : int }
+type 'a t = { head : 'a node Atomic.t; capacity : int }
```

Comment:
The `Cons capacity` field looks a bit redundant since it's only really useful near the root instead of at every level of the node list. Maybe `type 'a t = (int * 'a list) Atomic.t` would be easier? (with the `int` representing the room left in the stack, so when it reaches 0 we know it's full)

```ocaml
let capacity t = (* not needed, just to clarify the semantics *)
  let room_left, lst = Atomic.get t in room_left + List.length lst
```

But the O(1) `length` function was nice to have, so perhaps:

```ocaml
type 'a t = { head: (int * 'a list) Atomic.t ; capacity: int }
let length t = t.capacity - fst (Atomic.get t.head)
let is_full t = fst (Atomic.get t.head) = 0

(* or the fst could represent the list length: *)
let length t = fst (Atomic.get t.head)
let is_full t = t.capacity = length t
```

On a side-note, the `is_full` function could be nice to add to the API? :)

Reply:
Good point, thanks !